core/src/main/java/org/mule/processor/BlockingProcessorExecutor.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.processor;
+
+import org.mule.MessageExchangePattern;
+import org.mule.OptimizedRequestContext;
+import org.mule.VoidMuleEvent;
+import org.mule.api.MessagingException;
+import org.mule.api.MuleEvent;
+import org.mule.api.component.Component;
+import org.mule.api.endpoint.OutboundEndpoint;
+import org.mule.api.processor.MessageProcessor;
+import org.mule.api.processor.ProcessorExecutor;
+import org.mule.api.transformer.Transformer;
+import org.mule.execution.MessageProcessorExecutionTemplate;
+import org.mule.routing.MessageFilter;
+
+import java.util.List;
+
+/**
+ * This {@link org.mule.api.processor.ProcessorExecutor} implementation executes each {@link org.mule.api.processor.MessageProcessor}
+ * in sucession in the same thread until or processors have been invoked or one of the following is returned by a processor:
+ * <li>{@link org.mule.VoidMuleEvent}</li>
+ * <li><code>null</code></li>
+ */
+public class BlockingProcessorExecutor implements ProcessorExecutor
+{
+
+    protected final MessageProcessorExecutionTemplate messageProcessorExecutionTemplate;
+    protected final boolean copyOnVoidEvent;
+    protected final List<MessageProcessor> processors;
+
+    protected MuleEvent event;
+    private int index;
+
+    public BlockingProcessorExecutor(MuleEvent event, List<MessageProcessor> processors,
+                                     MessageProcessorExecutionTemplate messageProcessorExecutionTemplate, boolean
+            copyOnVoidEvent)
+    {
+        this.event = event;
+        this.processors = processors;
+        this.copyOnVoidEvent = copyOnVoidEvent;
+        this.messageProcessorExecutionTemplate = messageProcessorExecutionTemplate;
+    }
+
+    @Override
+    public MuleEvent execute() throws MessagingException
+    {
+        while (continueExecuting())
+        {
+            event = executeNext();
+        }
+        return event;
+    }
+
+    protected boolean continueExecuting()
+    {
+        return getIndex() < processors.size() && event != null && !(event instanceof VoidMuleEvent);
```

Comment:
Let's think about this today maybe, I don't think there will be a situation where two different threads are invoking continueExecuting() though.